### PR TITLE
Pytest arjun

### PR DESCRIPTION
**Pull request by ArjunPutcha**

### CODE REVIEW

File: mesmerize_napari/tests/test_core.py [hunk 1]
```diff
@@ -117,9 +118,9 @@ def test_mcorr():
     except:
         pytest.fail("Something wrong with setting UUID for batch items")
 
-    assert df.iloc[-1]['input_movie_path'] == os.path.join(vid_dir, f'{algo}.tif')
+    assert os.path.join(vid_dir,df.iloc[-1]['input_movie_path']) == \
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:121]: Let's import `get_full_data_path()` from `..core` and use that instead, so:

```python
assert get_full_data_path(df.iloc[-1]['input_movie_path']) == vid_dir.joinpath(f"{algo}.tif")
```

Comment (kushalkolar) [reply mesmerize_napari/tests/test_core.py:121]: probably useful to test this too to be fully pedantic:

```python
assert get_full_data_path(df.iloc[-1]['input_movie_path']) == vid_dir.joinpath(df.iloc[-1]['input_movie_path'])
```

File: mesmerize_napari/tests/test_core.py [hunk 2]
```diff
@@ -165,7 +167,8 @@ def test_cnmf():
     except:
         pytest.fail("Something wrong with setting UUID for batch items")
 
-    assert df.iloc[-1]['input_movie_path'] == os.path.join(vid_dir, f'{algo}.tif')
+    assert os.path.join(vid_dir, df.iloc[-1]['input_movie_path']) == \
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:170]: same comment as above

File: mesmerize_napari/tests/test_core.py [hunk 3]
```diff
@@ -191,7 +190,7 @@ def test_cnmf():
 
     algo = 'cnmf'
     print("Testing cnmf")
-    input_movie_path = df.iloc[-1]['outputs']['mcorr-output-path']
+    input_movie_path = os.path.join(vid_dir, df.iloc[-1]['outputs']['mcorr-output-path'])
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:193]: switch to using `pathlib`h

File: mesmerize_napari/tests/test_core.py [hunk 4]
```diff
@@ -208,7 +207,7 @@ def test_cnmf():
     except:
         pytest.fail("Something wrong with setting UUID for batch items")
     print('df input path:', df.iloc[-1]['input_movie_path'])
-    assert Path(df.iloc[-1]['input_movie_path']) == input_movie_path
+    assert os.path.join(vid_dir, df.iloc[-1]['input_movie_path']) == input_movie_path
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:210]: use `pathlib`

File: mesmerize_napari/tests/test_core.py [hunk 5]
```diff
@@ -218,7 +217,18 @@ def test_cnmf():
...
+    assert Path(os.path.join(vid_dir,
+        f'{df.iloc[-1]["uuid"]}_cnmf-memmap__d1_60_d2_80_d3_1_order_C_frames_2000_.mmap')) == \
+           get_full_data_path(df.iloc[-1]['outputs']['cnmf-memmap-path']) == \
+          Path(os.path.join(vid_dir, df.iloc[-1]['outputs']['cnmf-memmap-path']
+                       ))
+
+    assert Path(os.path.join(vid_dir,
+        f'{df.iloc[-1]["uuid"]}.hdf5')) == \
+        get_full_data_path(df.iloc[-1]['outputs']['cnmf-hdf5-path']) == \
+        Path(os.path.join(vid_dir, df.iloc[-1]['outputs']['cnmf-hdf5-path']))
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:231]: also test all the cnmf extensions to make sure the output is the same as what we get for the test dataset

**Review by kushalkolar — COMMENTED** [review]
In addition for line 19-20 use `pathlib` stuff instead of `os.path.join`, so basically make `tmp_dir` and `vid_dir` `Path` objects.

File: mesmerize_napari/tests/test_core.py [hunk 6]
```diff
@@ -73,8 +73,8 @@ def download_data(fname: str):
     return Path(path)
 
 
-def teardown_module():
-    clear_tmp()
+# def teardown_module():
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:76]: I'm assuming you're doing this to save time with tests while doing it locally? I think you can create an option to pass to pytest to not clear temp:
https://docs.pytest.org/en/6.2.x/example/parametrize.html

Create a file called `conftest.py`
```python
def pytest_addoption(parser):
    parser.addoption("--keep-test-data", action="store_true", help="keep test data after running tests")

def pytest_generate_tests(metafunc):
    if "keep_test_data" in metafunc.fixturenames:
        if metafunc.config.getoption("keep-test-data"):
            k = True
        else:
            k = False
        metafunc.parametrize("keep_test_data", k)
```

And in `test_core.py` make the `clear_tmp()` function like:

```python
def clear_tmp(keep_test_data):
    if keep_test_data:
        return
    else:
        # cleanup stuff
```

And then you can run pytest like this:

```
pytest --keep-test-data
```

Comment (ArjunPutcha) [reply mesmerize_napari/tests/test_core.py:76]: running pytest from terminal on laptop isn't working, I'm using pycharm ide. I'll do this when I test on nellab desktop

File: mesmerize_napari/tests/test_core.py (hunk 3 above)
Comment (ArjunPutcha) [reply mesmerize_napari/tests/test_core.py:193]: Great point!

File: mesmerize_napari/tests/test_core.py [hunk 7]
```diff
@@ -118,8 +118,10 @@ def test_mcorr():
     except:
         pytest.fail("Something wrong with setting UUID for batch items")
 
-    assert os.path.join(vid_dir,df.iloc[-1]['input_movie_path']) == \
-           os.path.join(vid_dir, f'{algo}.tif')
+    assert get_full_data_path(df.iloc[-1]['input_movie_path']) == \
+                  Path(vid_dir, f"{algo}.tif") == \
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:122]: no this won't work, change the stuff in line 19-20 to `Path` objects and then `Path.joinpath()`

https://docs.python.org/3/library/pathlib.html

File: mesmerize_napari/tests/test_core.py [hunk 8]
```diff
@@ -137,12 +139,12 @@ def test_mcorr():
         os.path.join(vid_dir,
         f'{df.iloc[-1]["uuid"]}-mcorr_els__d1_60_d2_80_d3_1_order_F_frames_2000_.mmap')
 
-    assert Path(os.path.join(vid_dir, df.iloc[-1]['outputs']['mcorr-output-path']
-                        )) == \
+    assert Path(vid_dir, df.iloc[-1]['outputs']['mcorr-output-path']
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:142]: see prevoius comment

File: mesmerize_napari/tests/test_core.py [hunk 9]
```diff
@@ -181,12 +183,12 @@ def test_cnmf():
     assert df.iloc[-1]['outputs']['success'] is True
     assert df.iloc[-1]['outputs']['traceback'] is None
 
-    assert Path(os.path.join(vid_dir, df.iloc[-1]['outputs']['mcorr-output-path']
-                        )) == \
+    assert Path(vid_dir, df.iloc[-1]['outputs']['mcorr-output-path']
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:186]: see previous comment

File: mesmerize_napari/tests/test_core.py [hunk 10]
```diff
@@ -231,22 +233,65 @@ def test_cnmf():
...
+    assert df.iloc[-1]['algo'] == algo
+    assert df.iloc[-1]['name'] == f'test-{algo}'
+    assert df.iloc[-1]['params'] == test_params[param_name]
+    assert df.iloc[-1]['outputs'] is None
+    try:
+        UUID(df.iloc[-1]['uuid'])
+    except:
+        pytest.fail("Something wrong with setting UUID for batch items")
+
+    assert os.path.join(vid_dir, df.iloc[-1]['input_movie_path']) == \
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:260]: `pathlib` this

File: mesmerize_napari/tests/test_core.py [hunk 11]
```diff
@@ -231,22 +233,65 @@ def test_cnmf():
...
+    process.wait()
+
+    df = load_batch(batch_path)
+    # Confirm output path is as expected
+    assert df.iloc[-1]['outputs']['success'] is True
+    assert df.iloc[-1]['outputs']['traceback'] is None
+    assert Path(vid_dir,
+                'memmap__d1_128_d2_128_d3_1_order_C_frames_1000_.mmap') == \
+           get_full_data_path(df.iloc[-1]['outputs']['cnmfe_memmap'])
+    assert Path(tmp_dir, f'{df.iloc[-1]["uuid"]}_pnr.pikl') == \
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:275]: see earlier comments

File: mesmerize_napari/tests/test_core.py [hunk 12]
```diff
@@ -262,21 +307,70 @@ def test_cnmfe():
...
+    assert df.iloc[-1]['name'] == f'test-{algo}'
+    assert df.iloc[-1]['params'] == test_params[algo]
+    assert df.iloc[-1]['outputs'] is None
+    try:
+        UUID(df.iloc[-1]['uuid'])
+    except:
+        pytest.fail("Something wrong with setting UUID for batch items")
 
+    assert get_full_data_path(df.iloc[-1]['input_movie_path']) == \
+                  Path(vid_dir, f"{algo}.tif") == \
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:343]: see earlier comments

File: mesmerize_napari/tests/test_core.py [hunk 13]
```diff
@@ -262,21 +307,70 @@ def test_cnmfe():
...
+    assert df.iloc[-1]['outputs'] is None
+    try:
+        UUID(df.iloc[-1]['uuid'])
+    except:
+        pytest.fail("Something wrong with setting UUID for batch items")
+
+    assert get_full_data_path(df.iloc[-1]['input_movie_path']) == \
+                  Path(vid_dir, f"{algo}.tif") == \
+                  Path(vid_dir, df.iloc[-1]['input_movie_path'])
+    # Check removing specific rows works
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:365]: useful to check that the paths for the batch item data also no longer exist

Comment (ArjunPutcha) [reply mesmerize_napari/tests/test_core.py:365]: Sorry can you explain this?

Comment (kushalkolar) [reply mesmerize_napari/tests/test_core.py:364]: When a batch item is deleted, check that any items related to that batch item (like mcorr output, cnmf output, corr imgs etc.) are deleted.

**Review by kushalkolar — COMMENTED** [review]
pathlib changes look good!

File: mesmerize_napari/tests/test_core.py (hunk 13 above)
Comment (ArjunPutcha) [reply mesmerize_napari/tests/test_core.py:364]: got it!

File: mesmerize_napari/tests/test_core.py [hunk 14]
```diff
@@ -31,7 +31,7 @@ def clear_tmp():
 
     test = os.listdir(vid_dir)
     for item in test:
-        if item.endswith(".npy") | item.endswith(".mmap") | item.endswith(".runfile"):
+        if item.endswith(".npy") | item.endswith(".mmap") | item.endswith(".runfile") | item.endswith(".hdf5"):
```
[the remark this answers is not in the document]
Comment (kushalkolar) [reply mesmerize_napari/tests/test_core.py:33]: @ArjunPutcha `clear_tmp()` needs to clear the entire temp dir. You could make an env variable to keep the files etc. to inspect things when working on your own system.

```python
if 'MESMERIZE_KEEP_TEST_DATA' in os.environ.keys():
    if not os.environ['MESMERIZE_KEEP_TEST_DATA'] = '1':
        os.remove('/path/to/test/files')
```

[the remark this answers is not in the document]
Comment (kushalkolar) [reply mesmerize_napari/tests/test_core.py:33]: And then when running pytest:

```bash
MESMERIZE_KEEP_TEST_DATA=1 pytest .
```

File: mesmerize_napari/tests/test_core.py [hunk 15]
```diff
@@ -13,44 +13,53 @@
...
+            os.remove(os.path.join(vid_dir, item))
+
+
+
+def get_datafile(fname: str):
+    local_path = Path(os.path.join(vid_dir, f'{fname}.tif'))
+    if local_path.is_file():
+        return local_path
+    else:
+        download_data(fname)
```
Comment (kushalkolar) [inline mesmerize_napari/tests/test_core.py:43]: This returns `None` and the tests fail, will fix now

**Review by kushalkolar — COMMENTED** [review]
Tests fail on first attempt, tmp dir not fully cleared

**Pull request merged**